File: web/src/types/webhook.ts
/**
 * A tenant's outbound webhook: a URL that receives signed POSTs when subscribed
 * events occur. The `secret` is returned only on create/reset (shown once).
 */
export interface Webhook {
  id: string
  tenant_id: string
  url: string
  secret?: string
  event_types: string[]
  description: string
  enabled: boolean
  created_by: string
  created_by_email: string
  created_at: string
}

/** One attempt-tracked delivery of an event to a webhook. */
export interface WebhookDelivery {
  id: string
  webhook_id: string
  event_type: string
  status: 'pending' | 'success' | 'failed'
  attempts: number
  response_status: number
  error?: string
  next_attempt_at: string
  created_at: string
  delivered_at?: string | null
}

export interface CreateWebhookInput {
  url: string
  event_types: string[]
  description?: string
}

/** Subscribes to every event. */
export const EVENT_TYPE_ALL = '*'

/** Curated events a webhook can subscribe to (besides "all"). */
export const WEBHOOK_EVENT_TYPES: { value: string; label: string }[] = [
  { value: 'flag.saved', label: 'Flag saved' },
  { value: 'flag.config.patched', label: 'Flag config changed' },
  { value: 'flag.deleted', label: 'Flag deleted' },
  { value: 'change.requested', label: 'Change requested' },
  { value: 'change.approved', label: 'Change approved' },
  { value: 'change.rejected', label: 'Change rejected' },
  { value: 'trigger.fired', label: 'Trigger fired' },
]

File: internal/models/webhook.go
package models

import (
	"encoding/json"
	"time"
)

// EventTypeAll subscribes a webhook to every event.
const EventTypeAll = "*"

// Webhook delivery statuses.
const (
	DeliveryPending = "pending"
	DeliverySuccess = "success"
	DeliveryFailed  = "failed"
)

// Webhook is a tenant's registered endpoint that receives signed POSTs when a
// subscribed event occurs. The secret signs the payload (HMAC-SHA256) and is
// returned to managers on create/reset only.
type Webhook struct {
	ID             string    `json:"id"`
	TenantID       string    `json:"tenant_id"`
	URL            string    `json:"url"`
	Secret         string    `json:"secret,omitempty"` // shown once, on create/reset
	EventTypes     []string  `json:"event_types"`
	Description    string    `json:"description"`
	Enabled        bool      `json:"enabled"`
	CreatedBy      string    `json:"created_by"`
	CreatedByEmail string    `json:"created_by_email"`
	CreatedAt      time.Time `json:"created_at"`
}

// Subscribes reports whether this webhook should receive the given event action.
func (w Webhook) Subscribes(action string) bool {
	for _, e := range w.EventTypes {
		if e == EventTypeAll || e == action {
			return true
		}
	}
	return false
}

// WebhookDelivery is one attempt-tracked delivery of an event to a webhook.
type WebhookDelivery struct {
	ID             string          `json:"id"`
	WebhookID      string          `json:"webhook_id"`
	EventType      string          `json:"event_type"`
	Payload        json.RawMessage `json:"payload"`
	Status         string          `json:"status"`
	Attempts       int             `json:"attempts"`
	ResponseStatus int             `json:"response_status"`
	Error          string          `json:"error,omitempty"`
	NextAttemptAt  time.Time       `json:"next_attempt_at"`
	CreatedAt      time.Time       `json:"created_at"`
	DeliveredAt    *time.Time      `json:"delivered_at,omitempty"`
}
